onnxruntime/core/providers/coreml/builders/model_builder.cc

```diff
@@ -390,13 +390,57 @@ void CreateEmptyFile(const std::string& filename) {
 
 #endif  // defined(COREML_ENABLE_MLPROGRAM)
 
-std::string GetModelOutputPath(bool create_ml_program) {
-  // path is used to create the ML Package directory for ML Program, and for the model directly otherwise.
-  auto path = util::GetTemporaryFilePath();
-  if (!create_ml_program) {
-    path += ".model.mlmodel";
-  }
+std::string GetModelOutputPath(const CoreMLOptions& coreml_options,
+                               const GraphViewer& graph_viewer) {
+  const std::string& subgraph_name = graph_viewer.Name();
+  std::string path;
+  if (coreml_options.ModelCacheDirectory().empty()) {
+    // path is used to create the ML Package directory for ML Program, and for the model directly otherwise.
+    path = util::GetTemporaryFilePath();
+    if (!coreml_options.CreateMLProgram()) {
+      path += ".model.mlmodel";
+    }
+  } else {
+    // subgraph_name is uniquely generated by
+    // onnxruntime/core/providers/coreml/coreml_execution_provider.cc::gen_metadef_name
+    // int metadef_id = metadef_id_generator_.GenerateId(graph_viewer, model_hash);
+    // MakeString(user_provide_key, "_", COREML, "_", model_hash, "_", metadef_id);
+    std::string_view cache_key = std::string_view(subgraph_name)
+                                     .substr(0, subgraph_name.find_first_of("_"));
+    // subgraph_short_name is metadef_id
+    std::string_view subgraph_short_name = std::string_view(subgraph_name)
+                                               .substr(subgraph_name.find_last_of("_") + 1);
+    path = MakeString(std::string(coreml_options.ModelCacheDirectory()), "/", cache_key);
+    ORT_THROW_IF_ERROR(Env::Default().CreateFolder(path));
```

Comment:
If we throw (ORT_THROW_IF_ERROR or ORT_ENFORCE) for file system errors the app MUST have logic to unset the cache. If it does not it is possibly completely dead until an update is deployed. The app developer may not have seen any file system errors in their testing so may not be aware of this requirement.

If we disable caching the app still works. The app user might report slow startup which the app developer can debug. If we have clearly logged the issue and that we are disabling caching it should be easy for them to do that.

I would rather not render an app useless pending an update which may take days or weeks to be available from the Apple App Store.

Reply:
I got you. But EP has to create folders whenever user enabled caching or not, is that right?